lib/puppet/provider/service/base.rb

```diff
@@ -30,17 +30,23 @@ def getps
   # parameter.
   def getpid
     @resource.fail "Either stop/status commands or a pattern must be specified" unless @resource[:pattern]
-    regex = Regexp.new(@resource[:pattern])
     ps = getps
+
     self.debug "Executing '#{ps}'"
-    IO.popen(ps) { |table|
-      table.each_line { |line|
-        if regex.match(line)
-          self.debug "Process matched: #{line}"
-          ary = line.sub(/^\s+/, '').split(/\s+/)
-          return ary[1]
-        end
-      }
+    table = Puppet::Util::Execution.execute(ps)
+
+    # Processes can add arbitrary sequences of bytes to the cmdline retrieved by
+    # 'ps -f', so drop encodings and match as binary.
+    regex = @resource[:pattern].dup
+    regex = Regexp.new(regex.force_encoding("ASCII-8BIT"))
+    table.force_encoding("ASCII-8BIT")
+
+    table.each_line { |line|
+      if regex.match(line)
+        self.debug "Process matched: #{line}"
+        ary = line.sub(/^\s+/, '').split(/\s+/)
```

Comment:
I'm not sure what the `\s` character class means in a `BINARY` encoding - I'm a little surprised that it even works to be honest, but I'm assuming it's looking at things byte by byte?.  A whitespace ASCII character in Ruby is a `[ \t\r\n\f]` - i.e. any of the following bytes:

* 0x20 (space)
* 0x09 (tab \t)
* 0x0D (carriage return \r)
* 0x0A (line feed \n)
* 0x0C (form feed \f)

I'm going to see if I can craft some Unicode data, that when treated as a `BINARY` (i.e. `ASCII-8BIT`) string, will not split properly with the above code.  When there are multibyte Unicode characters that have any of the above bytes, then I presume this fails - but I will validate for certain.

Further, I'm not sure what we should ever be using `ASCII-8BIT` on strings that we later intend to manipulate in any way, for this reason.  Reading and writing files directly as `BINARY` is usually the right way to handle things I think, but when parsing values.. but when concatenating / regex searching / etc... I'm not sure that it's correct.